sycl/test/basic_tests/implicit_device_copyable_types.cpp

```diff
@@ -0,0 +1,161 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+
+#include <sycl/sycl.hpp>
+#include <variant>
+
+struct ACopyable {
+  int i;
+  ACopyable() = default;
+  ACopyable(int _i) : i(_i) {}
+  ACopyable(const ACopyable &x) : i(x.i) {}
+};
+
+template <> struct sycl::is_device_copyable<ACopyable> : std::true_type {};
+
+int main() {
+  sycl::queue q;
+  {
+    std::pair<int, float> pair_arr[5];
+    std::pair<int, float> pair;
+    q.submit([&](sycl::handler &cgh) {
+       cgh.single_task([=]() {
+         std::pair<int, float> p0 = pair_arr[0];
+         std::pair<int, float> p = pair;
+       });
+     }).wait_and_throw();
+  }
+
+  {
+    std::pair<ACopyable, float> pair_arr[5];
+    std::pair<ACopyable, float> pair;
+    q.submit([&](sycl::handler &cgh) {
+       cgh.single_task([=]() {
+         std::pair<ACopyable, float> p0 = pair_arr[0];
+         std::pair<ACopyable, float> p = pair;
+       });
+     }).wait_and_throw();
+  }
+
+  {
+    std::tuple<int, float, bool> tuple_arr[5];
+    std::tuple<int, float, bool> tuple;
+    q.submit([&](sycl::handler &cgh) {
+       cgh.single_task([=]() {
+         std::tuple<int, float, bool> t0 = tuple_arr[0];
+         std::tuple<int, float, bool> t = tuple;
+       });
+     }).wait_and_throw();
+  }
+
+  {
+    std::tuple<ACopyable, float, bool> tuple_arr[5];
+    std::tuple<ACopyable, float, bool> tuple;
+    q.submit([&](sycl::handler &cgh) {
+       cgh.single_task([=]() {
+         std::tuple<ACopyable, float, bool> t0 = tuple_arr[0];
+         std::tuple<ACopyable, float, bool> t = tuple;
+       });
+     }).wait_and_throw();
+  }
+
+  {
+    std::variant<int, float, bool> variant_arr[5];
+    std::variant<int, float, bool> variant;
+    q.submit([&](sycl::handler &cgh) {
+       cgh.single_task([=]() {
+         std::variant<int, float, bool> v0 = variant_arr[0];
+         std::variant<int, float, bool> v = variant;
+       });
+     }).wait_and_throw();
+  }
+
+  {
+    std::variant<ACopyable> variant_arr[5];
+    std::variant<ACopyable> variant;
+    q.submit([&](sycl::handler &cgh) {
+       cgh.single_task([=]() {
+         // std::variant with complex types relies on virtual functions, so
+         // they cannot be used within sycl kernels
```

Comment:
Do we verify it anywhere? If not, I'd suggest to modify this test as

```
// RUN: compile and run
// RUN: compile-only -fsyntax-only -Xclang <enable error/warning verification in clang> -DCOMPILE_ONLY
...
#if COMPILE_ONLY
  {
    std::variant<ACopyable>
    // try to pass it to device
    // expected-error: ...
  }
#endif

Reply:
I developed this test and tried both in Linux and Windows. Turns out there's no error in Windows (I guess MSCV doesn't require virtual functions to implement `std::variant`), so I don't think we should have this test. Also, I will update my comment to say that this only happens in some implementations of `std::variant`.